# SQL Project

During the coronavirus pandemic apps for book lovers became very popular.

You've been given a database of one of the services competing in this market. It contains data on books, publishers, authors, and customer ratings and reviews of books. This information will be used to generate a value proposition for a new product.

Tasks:

* Find the number of books released after January 1, 2000.
* Find the number of user reviews and the average rating for each book.
* Identify the publisher that has released the greatest number of books with more than 50 pages (this will help you exclude brochures and similar publications from your analysis).
* Identify the author with the highest average book rating: look only at books with at least 50 ratings.
* Find the average number of text reviews among users who rated more than 50 books

# Table of Contents
1. **Step_1**:  
[Describe the goals of the study.](https://jupyterhub.praktikum-services.ru/user/user-0-1189703802/notebooks/13de20e2-005d-4d82-9574-678ceed7480c.ipynb#Describe-the-goals-of-the-study.)    

2. **Step_2**:  
[Connect to the database ](https://jupyterhub.praktikum-services.ru/user/user-0-1189703802/notebooks/13de20e2-005d-4d82-9574-678ceed7480c.ipynb#Connecting-to-the-data-base-and-uploading-data)   
       
3. **Step_3**:  
[Study the tables](https://jupyterhub.praktikum-services.ru/user/user-0-1189703802/notebooks/13de20e2-005d-4d82-9574-678ceed7480c.ipynb#Study-the-tables:)
    
4. **Step_4**:  
[Find the number of books released after January 1, 2000.](https://jupyterhub.praktikum-services.ru/user/user-0-1189703802/notebooks/13de20e2-005d-4d82-9574-678ceed7480c.ipynb#Find-the-number-of-books-released-after-January-1,-2000)  
        
5. **Step_5**:   
[Find the number of user reviews and the average rating for each book.](https://jupyterhub.praktikum-services.ru/user/user-0-1189703802/notebooks/13de20e2-005d-4d82-9574-678ceed7480c.ipynb#Find-the-number-of-user-reviews-and-the-average-rating-for-each-book.)

2. **Step_6**:  
[Identify the publisher that has released the greatest number of books with more than 50 pages](https://jupyterhub.praktikum-services.ru/user/user-0-1189703802/notebooks/13de20e2-005d-4d82-9574-678ceed7480c.ipynb#Identify-the-publisher-that-has-released-the-greatest-number-of-books-with-more-than-50-pages)   
       
3. **Step_7**:  
[Identify the author with the highest average book rating: look only at books with at least 50 ratings.](https://jupyterhub.praktikum-services.ru/user/user-0-1189703802/notebooks/13de20e2-005d-4d82-9574-678ceed7480c.ipynb#Identify-the-author-with-the-highest-average-book-rating-(look-only-at-books-with-at-least-50-ratings).)
    
4. **Step_8**:  
[Find the average number of text reviews among users who rated more than 50 books.](https://jupyterhub.praktikum-services.ru/user/user-0-1189703802/notebooks/13de20e2-005d-4d82-9574-678ceed7480c.ipynb#Find-the-average-number-of-text-reviews-among-users-who-rated-more-than-50-books.)  
        
5. **Step_5**:   
[Conclusion](https://jupyterhub.praktikum-services.ru/user/user-0-1189703802/notebooks/13de20e2-005d-4d82-9574-678ceed7480c.ipynb#Conclusion)

## Describe the goals of the study.

To create a  value proposition for a new product, we need to make a book market research using SQL in python: 
* Find the number of books released during last 20 years.
* Find the largest publisher.
* Find which type of books are the most popular.
* who is the top rated author.

## Connecting to the data base and uploading data

In [1]:
!pip install psycopg2-binary 

Defaulting to user installation because normal site-packages is not writeable


In [2]:
# import libraries
import pandas as pd
from sqlalchemy import create_engine


db_config = {'user': 'praktikum_student',         # user name
             'pwd': 'Sdf4$2;d-d30pp', # password
             'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
             'port': 6432,              # connection port
             'db': 'data-analyst-final-project-db'}          # the name of the data base

connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
                                                                     db_config['pwd'],
                                                                       db_config['host'],
                                                                       db_config['port'],
                                                                       db_config['db'])

engine = create_engine(connection_string, connect_args={'sslmode':'require'})

## Study the tables:

In [3]:
# creating an SQL query
query = '''
            SELECT * FROM books
        '''

In [4]:
books = pd.io.sql.read_sql(query, con = engine)
books.head()

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

In [5]:
query2 = '''
            SELECT * FROM authors
        '''

In [6]:
authors = pd.io.sql.read_sql(query2, con = engine)
authors.head()

author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

In [7]:
query3 = '''
            SELECT * FROM publishers
        '''

In [8]:
publishers = pd.io.sql.read_sql(query3, con = engine)
publishers.head()

publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

In [9]:
query4 = '''
            SELECT * FROM ratings
        '''

In [10]:
ratings = pd.io.sql.read_sql(query4, con = engine)
ratings.head()

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

In [11]:
query5 = '''
            SELECT * FROM reviews
        '''

In [12]:
reviews = pd.io.sql.read_sql(query5, con = engine)
reviews.head()

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

## Find the number of books released after January 1, 2000

In [13]:
number_of_books ='''
SELECT 
    COUNT(book_id) AS cnt
FROM 
    books
WHERE
    publication_date > '2000-01-01'
    
'''

In [14]:
number_of_books = pd.read_sql(number_of_books, engine)
number_of_books

cnt
0  819

There are 819 books which were released after January 1, 2000.

## Find the number of user reviews and the average rating for each book.

In [15]:
reviews_rating = '''
SELECT
    books.book_id AS book_id,
    books.title AS title,
    COUNT(DISTINCT reviews.review_id) AS reviews_number,
    AVG(ratings.rating) AS avg_rating
FROM
    books 
INNER JOIN reviews ON reviews.book_id = books.book_id
INNER JOIN ratings ON ratings.book_id = books.book_id
GROUP BY
    books.book_id
ORDER BY
    reviews_number DESC
LIMIT
    15;
    
'''

In [16]:
reviews_rating = pd.read_sql(reviews_rating, engine, index_col ='book_id')
reviews_rating

title  reviews_number  \
book_id                                                                      
948                                Twilight (Twilight  #1)               7   
854                                               The Road               6   
207                                        Eat  Pray  Love               6   
656                                         The Book Thief               6   
733                              The Giver (The Giver  #1)               6   
779      The Lightning Thief (Percy Jackson and the Oly...               6   
302      Harry Potter and the Prisoner of Azkaban (Harr...               6   
299      Harry Potter and the Chamber of Secrets (Harry...               6   
497                              Outlander (Outlander  #1)               6   
627                                          The Alchemist               6   
695      The Curious Incident of the Dog in the Night-Time               6   
696                 The Da Vinci Code (Robert Langdon  #2)               6   
734                                       The Glass Castle               6   
750                    The Hobbit  or There and Back Again               6   
673                                 The Catcher in the Rye               6   

         avg_rating  
book_id              
948        3.662500  
854        3.772727  
207        3.395833  
656        4.264151  
733        3.750000  
779        4.080645  
302        4.414634  
299        4.287500  
497        4.125000  
627        3.789474  
695        4.081081  
696        3.830508  
734        4.206897  
750        4.125000  
673        3.825581

The highest number of users reviews has the book "Twilight" with an average rating of 3.7.

## Identify the publisher that has released the greatest number of books with more than 50 pages

In [17]:
top_publishers = ''' 
SELECT 
    publishers.publisher,
    COUNT(DISTINCT book_id) AS books_count
FROM 
    books
LEFT JOIN 
    publishers 
ON 
    books.publisher_id = publishers.publisher_id
WHERE
    books.num_pages > '50'
GROUP BY 
    publishers.publisher_id
ORDER BY 
    books_count DESC 
LIMIT
    5
    '''

In [18]:
top_publishers = pd.read_sql(top_publishers, engine)
top_publishers

publisher  books_count
0             Penguin Books           42
1                   Vintage           31
2  Grand Central Publishing           25
3          Penguin Classics           24
4          Ballantine Books           19

Publisher "Penguin Books" has the greatest number of books released.

## Identify the author with the highest average book rating (look only at books with at least 50 ratings).

In [19]:
highest_rating='''
SELECT
    subquery.author AS author,
    AVG(subquery.avg) AS avg_rating
FROM
    (
    SELECT
        authors.author AS author,
        COUNT(ratings.rating) AS rating_cnt,
        AVG(ratings.rating) AS avg
    FROM
        ratings
    INNER JOIN books ON books.book_id = ratings.book_id
    INNER JOIN authors ON authors.author_id = books.author_id
    GROUP BY
        ratings.book_id,
        authors.author_id
    HAVING
        COUNT(ratings.rating)>50) AS subquery
GROUP BY
    author
ORDER BY
    avg_rating DESC;
'''

In [20]:
highest_rating = pd.read_sql(highest_rating, engine)
highest_rating

author  avg_rating
0                          J.K. Rowling/Mary GrandPré    4.283844
1                   Markus Zusak/Cao Xuân Việt Khương    4.264151
2                                      J.R.R. Tolkien    4.258446
3                                   Louisa May Alcott    4.192308
4                                        Rick Riordan    4.080645
5                                     William Golding    3.901408
6                                       J.D. Salinger    3.825581
7            Paulo Coelho/Alan R. Clarke/Özdemir İnce    3.789474
8   William Shakespeare/Paul Werstine/Barbara A. M...    3.787879
9                                           Dan Brown    3.754540
10                                         Lois Lowry    3.750000
11            George Orwell/Boris Grabnar/Peter Škerl    3.729730
12                                    Stephenie Meyer    3.662500
13                                     John Steinbeck    3.622951

J.K. Rowling is the author with the highest average book rating.

## Find the average number of text reviews among users who rated more than 50 books.

In [21]:
text_reviews_avg ='''
SELECT
    AVG(subquery.reviews_number) AS avg_reviews
FROM
    (SELECT  
        COUNT(DISTINCT reviews.review_id) AS reviews_number  
    FROM
        reviews
    LEFT JOIN ratings ON ratings.username = reviews.username
    GROUP BY
        reviews.username
    HAVING
        COUNT(DISTINCT ratings.rating_id) > 50) AS subquery;
'''

In [22]:
text_reviews_avg = pd.read_sql(text_reviews_avg, engine)
text_reviews_avg

avg_reviews
0    24.333333

On average users who rated more than 50 books wrote 24 text reviews.

## Conclusion

* For the last 20 years 819 books have been released. 
* The most talked about book became "Twilight", it has the biggest number of users reviews, but it's average rating is 3.7.
* The greatest publisher by the number of books released is Penguin Books. 
* J.K. Rowling is the author with the highest average book rating. 
* The most active users wrote 24 text reviews on average.